### PR TITLE
Text fetching

### PR DESCRIPTION
**Pull request by Fantaskink**
This pull request introduces a new `Fetcher` class for retrieving and processing book text from the Gutendex API, along with comprehensive tests for its functionality. It also adds the `requests` library as a dependency and updates the project structure to support these changes.

**Text fetching and processing functionality:**

* Added `Fetcher` class in `src/text_fetching/fetcher.py` that can fetch a random book's text from the Gutendex API, extract a random slice from the text, and format text to only lowercase alphabetic characters.

**Testing:**

* Added a suite of tests in `test/text-fetching/test_fetcher.py` to validate the `Fetcher` class methods, including edge cases for invalid input and text formatting.

**Dependency management:**

* Updated `pyproject.toml` to include `requests` as a required dependency for making HTTP requests.

**Project structure:**

* Modified `src/main.py` to remove the placeholder print statement, preparing it for future logic integration.

### CODE REVIEW

File: src/main.py
Comment (Jatewo) [inline src/main.py:1]: Perhaps cut down on the print use.

File: src/main.py [hunk 1]
```diff
@@ -1,6 +1,22 @@
+from text_fetching.fetcher import Fetcher
+
 def main():
-    print("Hello from ciphertext-generation!")
+	print("Hello from ciphertext-generation!")
```
Comment (Jatewo) [inline src/main.py]: Remove?

File: src/main.py [hunk 2]
```diff
@@ -1,6 +1,22 @@
...
+
+	fetcher = Fetcher()
+	book_text = fetcher.fetch_random_book_text()
+	if book_text:
+		print("Fetched book text successfully!")
+		print(f"First 500 characters:\n{book_text[:500]}")
```
Comment (Jatewo) [inline src/main.py]: Remove?

File: src/main.py [hunk 3]
```diff
@@ -1,6 +1,22 @@
...
+		slice_text = fetcher.get_random_book_slice(book_text)
+		if slice_text:
+			print(f"\nRandom slice ({len(slice_text)} characters):\n{slice_text}")
+			formatted = fetcher.format_text(slice_text)
+			print(f"\nFormatted text length (alphabetic chars only): {len(formatted)}")
+			print(f"First 100 characters of formatted text: {''.join(formatted[:100])}")
```
Comment (Jatewo) [inline src/main.py]: Remove?

File: src/main.py [hunk 4]
```diff
@@ -1,6 +1,22 @@
...
+	print("Hello from ciphertext-generation!")
+
+	fetcher = Fetcher()
+	book_text = fetcher.fetch_random_book_text()
+	if book_text:
+		print("Fetched book text successfully!")
+		print(f"First 500 characters:\n{book_text[:500]}")
+		slice_text = fetcher.get_random_book_slice(book_text)
+		if slice_text:
+			print(f"\nRandom slice ({len(slice_text)} characters):\n{slice_text}")
```
Comment (Jatewo) [inline src/main.py]: Remove?

File: src/text_fetching/fetcher.py
Comment (Jatewo) [inline src/text_fetching/fetcher.py:1]: I assume all the prints are for debugging. Perhaps remove?

File: src/text_fetching/fetcher.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			print(f"Downloading text from: {text_url}")
+			text_response = requests.get(text_url)
+			text_response.raise_for_status()
+			book_text = text_response.text
+			
+			print(f"Downloaded {len(book_text)} characters")
+			return book_text
+			
+		except requests.RequestException as e:
+			print(f"Error fetching book data: {e}")
```
Comment (Jatewo) [inline src/text_fetching/fetcher.py]: Perhaps raise an error here instead

File: src/text_fetching/fetcher.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		# Get random slice
+		start_idx = random.randint(0, max(0, len(book_text) - min_len))
+		end_idx = min(len(book_text), start_idx + random.randint(min_len, max_len))
+		slice_text = book_text[start_idx:end_idx]
+		
+		return slice_text
+		
+	def format_text(self, text: str) -> str:
+		if not isinstance(text, str):
+			raise ValueError(
```
Comment (Jatewo) [inline src/text_fetching/fetcher.py]: Unreachable code

File: src/text_fetching/fetcher.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+			
+			print(f"Downloaded {len(book_text)} characters")
+			return book_text
+			
+		except requests.RequestException as e:
+			print(f"Error fetching book data: {e}")
+			return None
+
+	def get_random_book_slice(self, book_text: str, min_len: int = 100, max_len: int = 5000) -> str | ValueError:
+		"""Extract a random slice from the provided book text."""
```
Comment (Jatewo) [inline src/text_fetching/fetcher.py]: Consider autoDocstring extension to help generate these descriptions (adds descriptions for arguments and return type as well)

File: test/text-fetching/test_fetcher.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    assert isinstance(formatted_text, str)
+    assert formatted_text == "imetatravellerfromanantiquelandwhosaidtwovastandtrunklesslegsofstonestandinthedesert"
+
+
+def test_format_notext(no_text):
+    fetcher = Fetcher()
+    with pytest.raises(ValueError) as excinfo:
+        fetcher.format_text(no_text)
+    assert (
+        "text must a string"
```
Comment (Mafusn) [inline test/text-fetching/test_fetcher.py]: "text must a string" 🍄

File: test/text-fetching/test_fetcher.py
Comment (Jatewo) [inline test/text-fetching/test_fetcher.py:1]: Add tests for remaining function

**Review by Jatewo — CHANGES_REQUESTED** [review]

**Review by Mafusn — APPROVED** [review]
LGTM 🦅

File: src/text_fetching/fetcher.py (hunk 6 above)
Comment (Fantaskink) [reply src/text_fetching/fetcher.py]: To my understanding, this is not unreachable, since python allows you to call the function with any type.

File: test/text-fetching/test_fetcher.py (hunk 8 above)
Comment (Fantaskink) [reply test/text-fetching/test_fetcher.py]: Fixed

File: src/text_fetching/fetcher.py (hunk 7 above)
Comment (Fantaskink) [reply src/text_fetching/fetcher.py]: Done

File: src/text_fetching/fetcher.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                    f"No suitable text format found for book ID {random_id}. Available formats: {list(formats.keys())}")
+
+            text_response = requests.get(text_url)
+            text_response.raise_for_status()
+            book_text = text_response.text
+
+            return book_text
+
+        except requests.RequestException as e:
+            print(f"Error fetching book data: {e}")
```
Comment (Jatewo) [inline src/text_fetching/fetcher.py]: Raise error

File: test/text-fetching/test_fetcher.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    )
+
+
+def test_slicing_text(long_text):
+    fetcher = Fetcher()
+    sliced_text = fetcher.get_random_book_slice(
+        book_text=long_text, min_len=100, max_len=5000)
+    assert len(sliced_text) <= 5000
+    assert len(sliced_text) >= 100
+    assert isinstance(sliced_text, str)
```
Comment (Jatewo) [inline test/text-fetching/test_fetcher.py]: This line should be above len or it might fail on len

**Review by Jatewo — CHANGES_REQUESTED** [review]

**Review by Jatewo — APPROVED** [review]
LGTM dawg 👍

**Review by Mafusn — APPROVED** [review]
LGTM 🌮
